src/serac/physics/contact/contact_data.cpp

```diff
@@ -0,0 +1,367 @@
+    mfem::Array<const mfem::Array<int>*> inactive_tdofs_vector(static_cast<int>(interactions_.size()));
+    int                                  inactive_tdofs_ct = 0;
+    for (int i{0}; i < inactive_tdofs_vector.Size(); ++i) {
+      inactive_tdofs_vector[i] = &interactions_[static_cast<size_t>(i)].inactiveDofs();
+      inactive_tdofs_ct += inactive_tdofs_vector[i]->Size();
+    }
+    auto             dof_offsets = pressureDofOffsets();
+    mfem::Array<int> inactive_tdofs(inactive_tdofs_ct);
+    inactive_tdofs_ct = 0;
+    for (int i{0}; i < inactive_tdofs_vector.Size(); ++i) {
+      if (inactive_tdofs_vector[i]) {
+        for (int d{0}; d < inactive_tdofs_vector[i]->Size(); ++d) {
+          inactive_tdofs[d + inactive_tdofs_ct] = (*inactive_tdofs_vector[i])[d] + dof_offsets[i];
+        }
+        inactive_tdofs_ct += inactive_tdofs_vector[i]->Size();
+      }
+    }
+    inactive_tdofs.GetMemory().SetHostPtrOwner(false);
+    mfem::Array<int> rows(numPressureDofs() + 1);
+    rows              = 0;
+    inactive_tdofs_ct = 0;
+    for (int i{0}; i < numPressureDofs(); ++i) {
+      if (inactive_tdofs_ct < inactive_tdofs.Size() && inactive_tdofs[inactive_tdofs_ct] == i) {
+        ++inactive_tdofs_ct;
+      }
+      rows[i + 1] = inactive_tdofs_ct;
+    }
+    rows.GetMemory().SetHostPtrOwner(false);
+    mfem::Vector ones(inactive_tdofs_ct);
+    ones = 1.0;
+    ones.GetMemory().SetHostPtrOwner(false);
+    mfem::SparseMatrix inactive_diag(rows.GetData(), inactive_tdofs.GetData(), ones.GetData(), numPressureDofs(),
+                                     numPressureDofs(), false, false, true);
+    // if the size of ones is zero, SparseMatrix creates its own memory which it
+    // owns.  explicitly prevent this...
+    inactive_diag.SetDataOwner(false);
+    auto& block_1_0 = static_cast<mfem::HypreParMatrix&>(block_J->GetBlock(1, 0));
+    auto  block_1_1 = new mfem::HypreParMatrix(block_1_0.GetComm(), block_1_0.GetGlobalNumRows(),
+                                              block_1_0.GetRowStarts(), &inactive_diag);
+    block_1_1->SetOwnerFlags(3, 3, 1);
+    block_J->SetBlock(1, 1, block_1_1);
+    // end building I_(inactive)
+  }
+  return block_J;
+}
+
+std::function<void(const mfem::Vector&, mfem::Vector&)> ContactData::residualFunction(
+    std::function<void(const mfem::Vector&, mfem::Vector&)> orig_r)
+{
+  return [this, orig_r](const mfem::Vector& u, mfem::Vector& r) {
+    const int disp_size = reference_nodes_->ParFESpace()->GetTrueVSize();
+
+    mfem::Vector u_blk, p_blk;
+    u_blk.MakeRef(const_cast<mfem::Vector&>(u), 0, disp_size);
+    p_blk.MakeRef(const_cast<mfem::Vector&>(u), disp_size, numPressureDofs());
+
+    mfem::Vector r_blk, g_blk;
+    r_blk.MakeRef(r, 0, disp_size);
+    g_blk.MakeRef(r, disp_size, numPressureDofs());
+
+    double dt = 1.0;
+    setDisplacements(u_blk);
+    // we need to call update first to update gaps
+    update(1, 1.0, dt);
```

Comment:
this function seems to do a lot? updates gaps, then updates other stuff after pressure is computed?  Can it be broken up into its different responsibilities and then named more concretely?

Reply:
Yes, but these changes will need to be made upstream in Tribol (see https://github.com/LLNL/Tribol/issues/19 there).  Once they have been made, the corresponding code in Serac will be updated in a future PR.